Tests/iaas/flavor-naming/flavor-add-extra-specs.py

```diff
@@ -270,22 +270,29 @@ def main(argv):
         names_to_check.extend(extra_names_to_check)
 
         # syntax check: compute flavorname instances
+        # Also select best match
+        bestln = 0
```

Comment:
I definitely agree that we have to be careful here. My suggestion would be as follows:

```python
collection = set()
# in the following, keep version to facilitate sorting afterwards
for flavorname in flavornames.values():
    name_v2 = outputter(flavorname)
    collection.add(("v1", SyntaxV1.from_v2(name_v2)))
    collection.add(("v2", name_v2))
    short_v2 = outputter(flavorname.shorten())
    # could check whether short_v2 != name_v2, but the set will swallow everything
    collection.add(("v1", SyntaxV1.from_v2(short_v2)))
    collection.add(("v2", short_v2))
names = [item[1] for item in sorted(collection, key=lambda item: item[0], -len(item[1]))]
for idx, name in enumerate(names):
    logger.debug(f"set scs:name-v{idx+1} to {name}")
```

Reply:
In prose: we derive the canonical name strings for each of the flavornames that we have parsed, and we collect those name strings (each one together with the version to facilitate sorting afterwards) in a set, so that we don't see any duplicates. Then we sort by version, then descending by length of the name. This way, we will arrive at our canonical order.

This approach will preserve all names and only make sure that shortened variants are present as well as both versions of the syntax.

Reply:
This approach doesn't use a reference at all, but we need one in order to derive the other extra_specs (cpu-type and disk0-type). Frankly I still think that these are hard to set IF the names don't agree. In this case, the user has to disambiguate, or we have to mandate some disambiguation rule; for instance: these extra_specs **may not underpromise** (is my opion and also what the standard requires for the standard SCS flavors), and in this case, we would compute the supremum of the flavornames (the highest promise in each category). Then I would also mandate that at least one of the flavornames MUST not underpromise.

Reply:
I would agree that the names must not contradict each other. If they do, we may throw an error and let the operator resolve it rather than second guessing what is intended.
But: As understatement is allowed (and in some situations even expected, if we for example expect provider to support nested virtualization `_hwv` but still recommend to not use the complicated flavor name *as primary name for it*, we do need to be prepared that we have several names, some not stating all capabilities.
Your approach is exactly right: We could look at all names and take the max that is being claimed anywhere as reality. (Only exception is where we have hard facts, such as the ram, vcpu and disk fields in the OpenStack nova anwers. Any claim beyond these is of course a lie that should result in a hard error and maybe -- protected by an option -- be fixed.)
My simple implementation was an approximation: I look at the flavor name with the most details, which will be the longest name. It ignores the case where you have several names that each describe some aspects of the flavor in more detail but no name that has it all.
If we find a good implementation (and it looks like you have the skills to do so while I do not without investing significant time) that does build the superset, that would be great. But if we document that this tool only does the right thing for several flavor names if at least one includes all capabilities, that would not be a serious limitation for its usefulness.

Reply:
> This approach doesn't use a reference at all, but we need one in order to derive the other extra_specs (cpu-type and disk0-type). 

I like the implementation, thanks!

> Frankly I still think that these are hard to set IF the names don't agree. In this case, the user has to disambiguate, or we have to mandate some disambiguation rule; for instance: these extra_specs **may not underpromise** (is my opion and also what the standard requires for the standard SCS flavors), and in this case, we would compute the supremum of the flavornames (the highest promise in each category). Then I would also mandate that at least one of the flavornames MUST not underpromise.

Thinking into the future: We will at some point stop mandating flavor names, which makes the names (and properties) in extra_specs our primary source of information to select the right flavor in an (SCS-compatible) operator-independent way. So we will have name-vN properties that do underpromise in order to match what users are searching for. Requiring at least one that contains all capabilities is fine and makes sense from my perspective. If it saves us some complexity in the code, even better!